rfcs/0000-rfc-forms.md

```diff
@@ -0,0 +1,148 @@
+
+### Non-Goals
+- The goal is not to provide a set of Validators 
+
+## Motivation
+
+In enterprise applications forms with complex validations often play a huge part. In React and Angular a lot of easy to use solutions exist (Formik, React Hook Form and Angular Reactive Forms). With web components several alternatives exist as well (@vaadin/form or https://github.com/realiarthur/lite-form), but none have really gained traction. It would be awesome to have lightweight standard in the Lit community or hopefully even beyond.  
+
+## Detailed Design
+
+This is heavily influenced by Angular Reactive Forms. A prototype has been implemented by @UserGalileo here: https://github.com/UserGalileo/lit-reactive-forms
+
+### General Overview
+In general there are three types of form elements:
+
+#### Form group: 
+A form group is the most high level concept. It is a group of form controls. It can also hold form arrays. It combines values and state into single objects. Form groups can be nested. It allows to run validation on all children. State of the children can also be set.  
+
+#### Form control: 
+This is the most low level concept. This model holds the value and state of an individual control. Changes to the model should be reflected in the rendered control. Validation can be run on this. 
+
+#### Form array: 
+This allows dynamic lists to be part of a form. They act as aggregators similar to a form group, but new children can be added or deleted at runtime. Form array children can either be form controls or form groups. (Form array would also be possible as a child, but not sure this would make sense, it might be a quick win API wise though.)   
+
+All of these should have listeners for changes to value and status, so that users can patch the values or status of dependant controls when necessary. 
+
+### Validation
+
+Validation should allow for async validators. All validators can be async by default Validators are pure functions.  There are two types of validators:
+
+#### Form control level validator
+
+This would be just a function declared as `(value) => string`. `value` being the current value of the control and the returned string being the potential error message describing the validation error. If an empty string is returned the control is valid. This is inspired by https://developer.mozilla.org/en-US/docs/Web/API/HTMLSelectElement/setCustomValidity
+
+Validators should only be run for controls that are connected. 
+
+#### Form group level validator
+
+This validator runs on a form group and gets passed all values of the children and returns an object of error messages for the children. 
+
+### How to define a form
+
+#### Defining the model 
+The model for a form can be created similar to Angular Reactive Forms through a `FormBuilder` class. Every control must be initialized with an initial value. The following example is taken from @UserGalilelo's prototype:
+
+````js
+fb = new FormBuilder(this);
+
+form = this.fb.group({
```

Comment:
What's the connection to `this`? Do these objects interact with the host in some way?

Reply:
@justinfagnani The host may use the Controller's properties to get information about a control, for example it could show in the template:
- its validity state
- its errors
- its UI state (touched, blurred, dirty...)

If some of these changes, the host should be notified in order to re-check the template